base/loading.jl

```diff
@@ -1762,23 +1828,32 @@ function compilecache(pkg::PkgId, path::String, internal_stderr::IO = stderr, in
     # write the checksum, _and then_ atomically move the file to `cachefile`.
     mkpath(cachepath)
     tmppath, tmpio = mktemp(cachepath)
+    tmppath_o, tmpio_o = mktemp(cachepath)
+    tmppath_so, tmpio_so = mktemp(cachepath)
     local p
     try
         close(tmpio)
-        p = create_expr_cache(pkg, path, tmppath, concrete_deps, internal_stderr, internal_stdout)
+        close(tmpio_o)
+        close(tmpio_so)
+        p = create_expr_cache(pkg, path, tmppath, tmppath_o, concrete_deps, internal_stderr, internal_stdout)
         if success(p)
+            # Run linker over tmppath_o
+            Linking.link_jilib(tmppath_o, tmppath_so)
+
+            # Read preferences hash back from .ji file (we can't precompute because
+            # we don't actually know what the list of compile-time preferences are without compiling)
+            prefs_hash = preferences_hash(tmppath)
+            cachefile = compilecache_path(pkg, prefs_hash)
+            ocachefile = string(chopsuffix(cachefile, ".ji"), ".", Base.Libc.dlext)
+
             # append checksum to the end of the .ji file:
             open(tmppath, "a+") do f
+                # TODO write path and checksum of ocachefile
```

Comment:
Can we just `DLL_EXPORT` `write_dependency_list` from `staticdata_utils.c` or is there something different needed?